### PR TITLE
feat: Apache avro serializer and Schema Registry support

### LINKED ISSUE #25: feat: support for Kafka schema binding and Avro serialization
We need to support Kafka schema binding and native serialization and compression.

### PR DESCRIPTION
**Pull request by dougolima**
# Description

Creates ApacheAvro serializer in a new package
Supports Schema Registry configuration on Cluster
Breaking Change: Change the IMessageSerializer and IMessageCompressor signatures to achieve better performance and supports Avro
Closes #25

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [ ] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: src/KafkaFlow/Producers/MessageProducer.cs [hunk 1]
```diff
@@ -117,6 +136,21 @@ await this.middlewareExecutor
 
                     return completionSource.Task;
                 });
+            }
+            catch (Exception ex) when (!(ex is ProduceException<byte[], byte[]> || ex is KafkaException))
+            {
+                this.dependencyResolverScope.Resolver
+                    .Resolve<ILogHandler>()
+                    .Error("Error executing consumer",
```
Comment (johnsonlu) [inline src/KafkaFlow/Producers/MessageProducer.cs:144]: `Error executing producer`?

Comment (dougolima) [reply src/KafkaFlow/Producers/MessageProducer.cs]: Thanks, fixed :)

**Review by zehenrique — COMMENTED** [review]
Left some feedback. Let me know if you have any question

File: src/KafkaFlow.Abstractions/Configuration/SchemaRegistryConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
+namespace KafkaFlow.Configuration
+{
+    /// <summary>
+    /// </summary>
+    public class SchemaRegistryConfiguration
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Abstractions/Configuration/SchemaRegistryConfiguration.cs:5]: Yes, if we use `SchemaRegistryConfig.cs` contained in the package `Confluent.SchemaRegistry` then we will need to reference `Confluent.SchemaRegistry` package in the KafkaFlow main project. In this case, everyone using Kafkaflow would have this reference as well, regardless of whether it is using schema registry or not.
The objective of KafkaFlow.Abstractions is to be this abstraction layer.

File: src/KafkaFlow.Abstractions/Configuration/AuthCredentialsSource.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+namespace KafkaFlow.Configuration
+{
+    /// <summary>AuthCredentialsSource enum values</summary>
+    public enum AuthCredentialsSource
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Abstractions/Configuration/AuthCredentialsSource.cs:4]: Same answers from [this](https://github.com/Farfetch/kafka-flow/pull/105#discussion_r566816259)

File: src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
+        <RootNamespace>KafkaFlow.Serializer.ApacheAvros</RootNamespace>
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj:5]: Yes, thanks. Fixed.

File: src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                        new AvroDeserializerConfig());
+            
+            return deserializer
+                .DeserializeAsync(data, data == null, SerializationContext.Empty)
+                .ConfigureAwait(false)
+                .GetAwaiter()
+                .GetResult();
+        }
+        
+        private void BuildSchemaRegistryClient(SchemaRegistryConfiguration schemaRegistryConfiguration)
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs:67]: Thanks for the suggestion, I did some changes according to your tips, can you take a look?

File: src/KafkaFlow.Abstractions/Configuration/IClusterConfigurationBuilder.cs [hunk 6]
```diff
@@ -25,6 +25,13 @@ public interface IClusterConfigurationBuilder
         /// <param name="handler">A handler to sets the values</param>
         /// <returns></returns>
         IClusterConfigurationBuilder WithSecurityInformation(Action<SecurityInformation> handler);
+        
+        /// <summary>
+        /// Configures schema registry
+        /// </summary>
+        /// <param name="configuration">Schema registry configuration</param>
+        /// <returns></returns>
+        IClusterConfigurationBuilder WithSchemaRegistry(Action<SchemaRegistryConfiguration> configuration);
```
Comment (filipeesch) [inline src/KafkaFlow.Abstractions/Configuration/IClusterConfigurationBuilder.cs]: If you move this method to KafkaFlow package as an extension method, the implementation will be easier.

File: src/KafkaFlow.Abstractions/Configuration/SchemaRegistryConfiguration.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Abstractions/Configuration/SchemaRegistryConfiguration.cs:5]: Nevermind, based on [this ](https://github.com/Farfetch/kafka-flow/pull/105#discussion_r572175281), this is file is not needed anymore. :)

File: src/KafkaFlow.Abstractions/IMessageSerializer.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 namespace KafkaFlow
 {
     using System;
+    using Configuration;
```
Comment (filipeesch) [inline src/KafkaFlow.Abstractions/IMessageSerializer.cs:4]: undo this

File: src/KafkaFlow.Microsoft.DependencyInjection/MicrosoftDependencyResolver.cs [hunk 8]
```diff
@@ -15,7 +15,7 @@ public MicrosoftDependencyResolver(IServiceProvider serviceProvider)
 
         public object Resolve(Type type)
         {
-            return this.serviceProvider.GetRequiredService(type);
+            return this.serviceProvider.GetService(type);
```
Comment (filipeesch) [inline src/KafkaFlow.Microsoft.DependencyInjection/MicrosoftDependencyResolver.cs:18]: why this change?

File: src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    using Avro.Specific;
+    using Configuration;
+    using Confluent.Kafka;
+    using Confluent.Kafka.SyncOverAsync;
+    using Confluent.SchemaRegistry;
+    using Confluent.SchemaRegistry.Serdes;
+    /// <summary>
+    /// A message serializer using Apache.Avro library
+    /// </summary>
+    public class ApacheAvroMessageSerializer : IMessageSerializer
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs:13]: Thanks for the suggestion, we will think about it in the next releases.

File: src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (zehenrique) [reply src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs]: looks good

**Review by zehenrique — APPROVED** [review]

File: src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    <ItemGroup>
+      <ProjectReference Include="..\KafkaFlow.Abstractions\KafkaFlow.Abstractions.csproj" />
+      <ProjectReference Include="..\KafkaFlow.Serializer\KafkaFlow.Serializer.csproj" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <PackageReference Include="Apache.Avro" Version="1.10.1" />
+      <PackageReference Include="Confluent.Kafka" Version="1.5.3" />
+      <PackageReference Include="Confluent.SchemaRegistry" Version="1.5.3" />
+      <PackageReference Include="Confluent.SchemaRegistry.Serdes.Avro" Version="1.5.3" />
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj:16]: should use the same Confluent version of other projects

File: src/KafkaFlow.Serializer.Json/JsonMessageSerializer.cs [hunk 11]
```diff
@@ -3,6 +3,7 @@
     using System;
     using System.Text;
     using System.Text.Json;
+    using Configuration;
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.Json/JsonMessageSerializer.cs:6]: undo this

File: src/KafkaFlow.Serializer.NewtonsoftJson/NewtonsoftJsonMessageSerializer.cs [hunk 12]
```diff
@@ -2,6 +2,7 @@
 {
     using System;
     using System.Text;
+    using Configuration;
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.NewtonsoftJson/NewtonsoftJsonMessageSerializer.cs:5]: undo

File: src/KafkaFlow.Serializer/SerializerConsumerMiddleware.cs [hunk 13]
```diff
@@ -2,6 +2,7 @@
 {
     using System;
     using System.Threading.Tasks;
+    using Configuration;
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer/SerializerConsumerMiddleware.cs:5]: undo

File: src/KafkaFlow/Producers/MessageProducer.cs [hunk 14]
```diff
@@ -44,20 +44,37 @@ internal class MessageProducer : IMessageProducer, IDisposable
...
+                    .Error("Error executing producer",
+                        ex,
+                        new
+                        {
+                            message,
+                            topic,
+                            partitionKey
+                        });
+                throw;
+            }
```
Comment (filipeesch) [inline src/KafkaFlow/Producers/MessageProducer.cs:77]: I think we should not change this behaviour. The app is responsible to catch the exceptions and choose whats to do in these cases.

File: src/KafkaFlow/Producers/MessageProducer.cs [hunk 15]
```diff
@@ -117,6 +136,21 @@ await this.middlewareExecutor
...
```
Comment (filipeesch) [inline src/KafkaFlow/Producers/MessageProducer.cs:153]: same as other

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow.Abstractions/IMessageSerializer.cs (hunk 7 above)
Comment (dougolima) [reply src/KafkaFlow.Abstractions/IMessageSerializer.cs:4]: done

File: src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj (hunk 10 above)
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj:16]: Confluent package was updated to version 1.5.3

File: src/KafkaFlow.Serializer.Json/JsonMessageSerializer.cs (hunk 11 above)
Comment (dougolima) [reply src/KafkaFlow.Serializer.Json/JsonMessageSerializer.cs:6]: done

File: src/KafkaFlow.Serializer.NewtonsoftJson/NewtonsoftJsonMessageSerializer.cs (hunk 12 above)
Comment (dougolima) [reply src/KafkaFlow.Serializer.NewtonsoftJson/NewtonsoftJsonMessageSerializer.cs:5]: done

File: src/KafkaFlow.Serializer/SerializerConsumerMiddleware.cs (hunk 13 above)
Comment (dougolima) [reply src/KafkaFlow.Serializer/SerializerConsumerMiddleware.cs:5]: done

File: src/KafkaFlow/Producers/MessageProducer.cs (hunk 14 above)
Comment (dougolima) [reply src/KafkaFlow/Producers/MessageProducer.cs:77]: done

File: src/KafkaFlow/Producers/MessageProducer.cs (hunk 15 above)
Comment (dougolima) [reply src/KafkaFlow/Producers/MessageProducer.cs:153]: done

File: src/KafkaFlow.Microsoft.DependencyInjection/MicrosoftDependencyResolver.cs (hunk 8 above)
Comment (dougolima) [reply src/KafkaFlow.Microsoft.DependencyInjection/MicrosoftDependencyResolver.cs:18]: When a type that was not registered is requested, `GetRequiredService` throws an exception, whilst `GetService` returns null.
`AddApacheAvroSerializer` extensions need to try to resolve `ISchemaRegistryClient` and send the instance (or null) to `ApacheAvroMessageSerializer`. 
If there is no `ISchemaRegistryClient` instance, then `ApacheAvroMessageSerializer` will thrown a proper exception.

Comment (dougolima) [reply src/KafkaFlow.Microsoft.DependencyInjection/MicrosoftDependencyResolver.cs:18]: Another approach would be to keep `GetRequiredService` and use a try/catch block when trying to get the `ISchemaRegistryClient` instance in the `AddApacheAvroSerializer` extension. 
IMHO, it would be the best approach, what do you think?

File: src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+            AvroSerializerConfig serializerConfig)
+        {
+            this.schemaRegistryClient = schemaRegistryClient ?? throw new InvalidOperationException(NoSchemaRegistryUrlMessage);
+            this.serializerConfig = serializerConfig ;
+        }
+
+        /// <inheritdoc/>
+        public byte[] Serialize(object message)
+        {
+            return new AvroSerializer<ISpecificRecord>(
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs:43]: done

File: src/KafkaFlow.Serializer.ApacheAvro/ConfigurationExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
+namespace KafkaFlow.Serializer.ApacheAvro
+{
+    using System;
+    using Configuration;
+    using Confluent.SchemaRegistry;
+    using Confluent.SchemaRegistry.Serdes;
+    
+    public static class ClusterConfigurationBuilderExtensions
+    {
+        public static IClusterConfigurationBuilder WithSchemaRegistry(
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.ApacheAvro/ConfigurationExtensions.cs:10]: missing summary description, in public methods it is a must have 😃

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        /// <summary>
+        /// </summary>
+        /// <param name="resolver">The <see cref="IDependencyResolver"/> to be used by the framework</param>
+        /// <param name="serializerConfig">Avro serializer configuration</param>
+        public ApacheAvroMessageSerializer( 
+            IDependencyResolver resolver,
+            AvroSerializerConfig serializerConfig)
+        {
+            this.schemaRegistryClient = resolver.Resolve<ISchemaRegistryClient>() ?? 
+                                        throw new InvalidOperationException("There was not defined a schema registry to be used by ApacheAvroSerializer. Use 'WithSchemaRegistry' option at cluster level with a valid url.");
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs:35]: done

File: src/KafkaFlow.Serializer.ApacheAvro/ConfigurationExtensions.cs (hunk 17 above)
Comment (dougolima) [reply src/KafkaFlow.Serializer.ApacheAvro/ConfigurationExtensions.cs:10]: done

**Review by filipeesch — APPROVED** [review]

File: src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj [hunk 19]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
+        <PackageIconUrl>https://raw.githubusercontent.com/Farfetch/.github/master/images/fuse-logo-128.png</PackageIconUrl>
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.ApacheAvro/KafkaFlow.Serializer.ApacheAvro.csproj:5]: missing packageId and description

**Review by filipeesch — CHANGES_REQUESTED** [review]

**Review by filipeesch — APPROVED** [review]
